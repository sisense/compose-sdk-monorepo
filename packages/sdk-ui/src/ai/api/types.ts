import { JaqlQueryPayload } from '@sisense/sdk-query-client';

import { ExpandedQueryModel } from '@/analytics-composer';
import { WidgetProps } from '@/props';

export interface ChatContext {
  title: string;
  live: boolean;
}

export interface ChatContextDetails {
  dashboardId?: string;
}
export interface TextMessage {
  role: 'assistant';
  content: string; // stringified json { answer: string }
  type: 'text';
}
export interface NlqMessage {
  role: 'assistant';
  content: string; // stringified json
  type: 'nlq';
}
interface RegularMessage {
  role: 'user' | 'assistant';
  content: string;
}
export type ChatMessage = RegularMessage | NlqMessage | TextMessage;

export interface Chat {
  chatId: string;
  contextId: string;
  contextTitle: string;
  contextDetails?: ChatContextDetails;
  expireAt: string;
  tenantId: string;
  userId: string;
  chatHistory: ChatMessage[];
}
export type ChatWithoutHistory = Omit<Chat, 'chatHistory'>;

export interface LlmConfig {
  provider: string;
  apiKey: string;
  baseUrl: string;
  model: string;
  version?: string;
}

export interface ChatRequest {
  text?: string;
  timezone?: string;
  options?: {
    enableFollowup?: boolean;
  };
}

interface TextResponse {
  chatHistory: ChatMessage[];
  data: {
    answer: string;
  };
  chatId: string;
  responseType: 'text';
}

interface ErrorResponse {
  data: {
    message: string;
    code: string;
    answer: string;
  };
  responseType: 'error';
}

/**
 * Response data for NLQ request
 */
export interface NlqResponseData extends ExpandedQueryModel {
  /** Detailed description of the response */
  detailedDescription: string;
  /** An array of suggested followup questions  */
  followupQuestions?: string[];
  /** NLQ prompt used in the request  */
  nlqPrompt: string;
  /** Suggested widget props*/
  widgetProps?: WidgetProps | undefined;
  /** The response message for the chat  */
  userMsg: string;
  /** @internal */
  clarification?: string;
}
export interface NlqResponse {
  data: NlqResponseData;
  timestamp: string;
  responseType: 'nlq';
}
export type ChatResponse = NlqResponse | TextResponse | ErrorResponse;

export interface QueryRecommendationConfig {
  numOfRecommendations: number;
  userPrompt?: string;
}

export type NlqResult = Omit<NlqResponseData, 'followupQuestions'>;

/** AI-generated recommended query you can run on your data model */
export interface QueryRecommendation extends Omit<NlqResponseData, 'followupQuestions'> {}
export type QueryRecommendationResponse = QueryRecommendation[];

export interface GetNlgInsightsRequest {
  jaql: Pick<JaqlQueryPayload, 'datasource' | 'metadata' | 'filterRelations'>;
  verbosity?: 'Low' | 'High';
}

/** @internal */
export interface GetNlgInsightsResponse {
  data?: {
    answer: string;
  };
  responseType: 'Text';
}

/** @internal */
export const NLQ_RESULT_CHART_TYPES = [
  'indicator',
  'column',
  'line',
  'table',
  'pie',
  'bar',
] as const;
/** NLQ result chart types @expandType */
export type NlqResultChartType = (typeof NLQ_RESULT_CHART_TYPES)[number];
export interface GetNlqResultRequest {
  text: string;
  timezone?: string;
  chartTypes: NlqResultChartType[];
}

export interface SendFeedbackRequest {
  type: string;
  data: object;
  sourceId: string;
  rating: -1 | 1;
}

/**
 * The chat mode to use for a chat session
 *
 * analyze: Enable business users to uncover data insights
 * develop: Enable developers to build queries and charts for embedded analytics
 */
export type ChatMode = 'analyze' | 'develop';
