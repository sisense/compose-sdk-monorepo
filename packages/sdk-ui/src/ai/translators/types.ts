import { MetadataItem } from '@sisense/sdk-query-client';
import { AxesMapping, NlqResponseData } from '@/ai';
import { ChartType, TableType } from '@/types';
import { WidgetModel } from '@/models';

/**
 * Expanded Query Model that is based on NlqResponseData.
 * It contains expanded JAQL and chart recommendations returned from the chat response.
 *
 * @internal
 */
export type ExpandedQueryModel = Pick<
  NlqResponseData,
  'jaql' | 'chartRecommendations' | 'queryTitle'
>;

/**
 * Expanded Query Model
 *
 * Used as a default value for ExpandedQueryModel
 *
 * @internal
 */
export const EMPTY_EXPANDED_QUERY_MODEL: ExpandedQueryModel = {
  jaql: { datasource: { title: '' }, metadata: [] },
  queryTitle: '',
  chartRecommendations: {},
};

/**
 * Simple Chart Recommendations
 *
 * @internal
 */
export interface SimpleChartRecommendations {
  chartType: string;
  dataOptions: AxesMapping;
}

/**
 * Simple Query Model that is a simplified version of ExpandedQueryModel.
 *
 * It removes any JAQL syntax or information that already exists in the schema.
 * Also, any parameters that have a default do not need to be specified.
 *
 * Users manipulates this model via the query in YAML format.
 *
 * ModelTranslator is responsible for translating this query model to SimpleWidgetModel.
 *
 * @internal
 */
export interface SimpleQueryModel {
  /**
   * The data model or perspective title.
   */
  model: string;
  metadata: MetadataItem[];
  chart: SimpleChartRecommendations | {};
  /**
   * Title of the query
   */
  queryTitle: string;
}

/**
 * Empty Simple Query Model
 *
 * Used as a default value for SimpleQueryModel
 *
 * @internal
 */
export const EMPTY_SIMPLE_QUERY_MODEL: SimpleQueryModel = {
  model: '',
  metadata: [],
  chart: {},
  queryTitle: '',
};

/**
 * All Chart Types including Table
 * Note: Table will be re-united with ChartType in a future MR
 *
 * @internal
 */
export type AllChartType = ChartType | TableType;

/**
 * Chart Widget Model
 *
 * This model serves two purposes:
 *
 * (1) For rendering a chart or table widget in the Query Composer.
 * (2) For reverse engineered (translated) to CSDK code
 *
 * Naming TBD
 *
 * @internal
 */
export type ChartWidgetModel = Pick<WidgetModel, 'getChartWidgetProps' | 'getTableWidgetProps'> & {
  chartType: AllChartType;
  queryTitle: string;
};

/**
 * UI Framework
 *
 * @internal
 */
export type UiFramework = 'react' | 'vue' | 'angular';

/**
 * Code Template Key
 *
 * @internal
 */
export type CodeTemplateKey = 'baseChartTmpl' | 'chartTmpl' | 'chartWidgetTmpl' | 'tableWidgetTmpl';

/**
 * Code Templates
 *
 * @internal
 */
export type CodeTemplates = {
  [key in UiFramework]: { [key in CodeTemplateKey]: string };
};
