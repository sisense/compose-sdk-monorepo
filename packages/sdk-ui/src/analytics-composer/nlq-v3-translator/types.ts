import { Filter, FilterRelations, JSONArray, JSONValue, Measure } from '@sisense/sdk-data';

import { DataSchemaContext, NlqTranslationInput } from '../types';

export const ATTRIBUTE_PREFIX = 'DM.';

// Argument types for function validation
export type ArgType =
  // Basic types
  | 'string' // "Total Revenue", "DESC", etc.
  | 'string[]' // ["2012-01-01T00:00:00"], member arrays
  | 'number' // 5, 100, ranking values
  | 'number[]' // [10, 20] for numeric ranges
  | 'boolean' // true/false flags

  // SDK Object types (from function calls)
  | 'Attribute' // "DM.Commerce.Revenue" → Attribute
  | 'Attribute[]' // ["DM.Commerce.AgeRange"] → Attribute[]
  | 'Measure' // Function call → Measure (general)
  | 'BaseMeasure' // Function call → BaseMeasure (specific aggregation)
  | 'Measure[]' // Array of function calls → Measure[]
  | 'Filter' // Function call → Filter
  | 'Filter[]' // Array of function calls → Filter[]
  | 'FilterRelationsNode' // Filter | FilterRelations | Filter[]
  | 'DateDimension' // Date-specific attributes
  | 'LevelAttribute' // Date level attributes

  // Union types
  | 'Measure | number' // Can be either measure function or number
  | 'Date | string' // Date objects or ISO date strings
  | 'string | number' // String or numeric values

  // Configuration objects
  | 'BaseFilterConfig' // Optional filter config
  | 'MembersFilterConfig' // Members-specific filter config
  | 'CustomFormulaContext' // Custom formula context object

  // Catch-all for complex objects
  | 'any'; // Skip validation for complex types

// Argument schema for function validation
export interface ArgSchema {
  type: ArgType;
  required: boolean;
}

// Type to represent a function argument
export type Arg = string | number | FunctionCall | string[];
export type ProcessedArg = unknown;
export type QueryElement = Filter | FilterRelations | Measure;
export type FactoryFunction = (...args: unknown[]) => QueryElement;

// Type to represent the structure parsed from the JSON
export type FunctionCall = {
  function: string;
  args: Arg[];
};

/**
 * Context object passed to all custom processors containing runtime information
 * needed for processing.
 */
export interface FunctionContext extends DataSchemaContext {
  /** Path prefix for error messages (e.g., 'args[0].function') */
  pathPrefix: string;
}

export interface ArgContext extends FunctionContext {
  argSchema: ArgSchema;
}

// Translation input types
export type DimensionsInput = NlqTranslationInput<JSONArray, DataSchemaContext>;
export type MeasuresInput = NlqTranslationInput<JSONArray, DataSchemaContext>;
export type FiltersInput = NlqTranslationInput<JSONArray, DataSchemaContext>;
export type HighlightsInput = NlqTranslationInput<JSONArray, DataSchemaContext>;

// Function call input types
export type MeasuresFunctionCallInput = NlqTranslationInput<FunctionCall[], DataSchemaContext>;
export type FiltersFunctionCallInput = NlqTranslationInput<FunctionCall[], DataSchemaContext>;
export type HighlightsFunctionCallInput = NlqTranslationInput<FunctionCall[], DataSchemaContext>;

// Core processing input types
export type NodeInput = NlqTranslationInput<FunctionCall, FunctionContext>;
export type ArgInput = NlqTranslationInput<Arg, ArgContext>;

/**
 * Custom processor function signature.
 */
export type CustomFunctionProcessor = (
  processedArgs: ProcessedArg[],
  context: FunctionContext,
) => void;

export function isFunctionCall(value: JSONValue): value is FunctionCall {
  return typeof value === 'object' && value !== null && 'function' in value && 'args' in value;
}

export function isFunctionCallArray(value: JSONArray): value is FunctionCall[] {
  return value.every(isFunctionCall);
}

// Type guard to check if the argument is a Filter
export function isFilterElement(arg: QueryElement): arg is Filter {
  return 'attribute' in arg && 'config' in arg;
}

// Type guard to check if the argument is a FilterRelations
export function isFilterRelationsElement(arg: QueryElement): arg is FilterRelations {
  return 'left' in arg && 'right' in arg && 'operator' in arg;
}

export function isMeasureElement(arg: QueryElement): arg is Measure {
  return !isFilterElement(arg) && !isFilterRelationsElement(arg);
}

export function isStringArray(value: JSONArray): value is string[] {
  return value.every((item) => typeof item === 'string');
}
