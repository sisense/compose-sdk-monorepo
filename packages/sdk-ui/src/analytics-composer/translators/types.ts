import { MetadataItem } from '@sisense/sdk-query-client';
import { AxesMapping, NlqResponseData } from '../../ai';
import { ChartType, WidgetModel } from '../../';

/**
 * Expanded Query Model that is based on NlqResponseData.
 * It contains expanded JAQL and chart recommendations returned from the chat response.
 *
 * @internal
 */
export type ExpandedQueryModel = Pick<
  NlqResponseData,
  'jaql' | 'chartRecommendations' | 'queryTitle'
>;

/**
 * Expanded Query Model
 *
 * Used as a default value for ExpandedQueryModel
 *
 * @internal
 */
export const EMPTY_EXPANDED_QUERY_MODEL: ExpandedQueryModel = {
  jaql: { datasource: { title: '' }, metadata: [] },
  queryTitle: '',
  chartRecommendations: {},
};

/**
 * Simple Chart Recommendations
 *
 * @internal
 */
export interface SimpleChartRecommendations {
  chartType: string;
  dataOptions: AxesMapping;
}

/**
 * Simple Query Model that is a simplified version of ExpandedQueryModel.
 *
 * It removes any JAQL syntax or information that already exists in the schema.
 * Also, any parameters that have a default do not need to be specified.
 *
 * Users manipulates this model via the query in YAML format.
 *
 * ModelTranslator is responsible for translating this query model to SimpleWidgetModel.
 *
 * @internal
 */
export interface SimpleQueryModel {
  /**
   * The data model or perspective title.
   */
  model: string;
  metadata: MetadataItem[];
  chart: SimpleChartRecommendations | {};
  /**
   * Title of the query
   */
  queryTitle: string;
}

/**
 * Empty Simple Query Model
 *
 * Used as a default value for SimpleQueryModel
 *
 * @internal
 */
export const EMPTY_SIMPLE_QUERY_MODEL: SimpleQueryModel = {
  model: '',
  metadata: [],
  chart: {},
  queryTitle: '',
};

/**
 * Chart Widget Model
 *
 * This model serves two purposes:
 *
 * (1) For rendering a chart widget (including table) in the Query Composer.
 * (2) For reverse engineered (translated) to CSDK code
 *
 * @internal
 */
export type ChartWidgetModel = Pick<WidgetModel, 'getChartWidgetProps'> & {
  chartType: ChartType;
  queryTitle: string;
};

/**
 * UI Framework
 *
 * @internal
 */
export type UiFramework = 'react' | 'vue' | 'angular';

/**
 * Code Template Key
 *
 * @internal
 */
export type CodeTemplateKey = 'baseChartTmpl' | 'chartTmpl' | 'chartWidgetTmpl';

/**
 * Code Templates
 *
 * @internal
 */
export type CodeTemplates = {
  [key in UiFramework]: { [key in CodeTemplateKey]: string };
};
