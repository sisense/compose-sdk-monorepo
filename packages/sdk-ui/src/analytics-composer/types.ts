import { JaqlDataSourceForDto, JSONArray, JSONValue, MetadataItem } from '@sisense/sdk-data';

import { AnyColumn } from '@/chart-data-options/types.js';
import { DashboardProps } from '@/dashboard/types.js';
import { ChartWidgetProps, PivotTableWidgetProps, WidgetProps } from '@/props';
import { ExecutePivotQueryParams, ExecuteQueryParams } from '@/query-execution';
import { ChartDataOptions, ChartStyleOptions } from '@/types';

import { DynamicChartType } from '../chart-options-processor/translations/types.js';

type Stringify<T> = {
  [K in keyof T as `${K & string}String`]: string;
};

type ExtraWidgetCodeProps = {
  componentString: string;
  extraImportsString: string;
  idString?: string;
  widgetTypeString?: 'chart' | 'pivot' | 'text' | 'custom';
};

/** @internal */
export type KeysOfUnion<T> = T extends T ? keyof T : never;
/** @internal */
export type AllPossibleChartOptionKeys = KeysOfUnion<ChartDataOptions>;
/** @internal */
export type AxesMappingKey = Exclude<AllPossibleChartOptionKeys, 'seriesToColorMap'>;
/** @internal */
export type AxesMapping = Partial<Record<AxesMappingKey, Array<AnyColumn>>>;

/** @internal */
export interface ChartRecommendations {
  chartFamily: string;
  chartType: string;
  axesMapping: AxesMapping;
  styleOptions?: ChartStyleOptions;
}

/**
 * Expanded Query Model that is based on NlqResponseData.
 * It contains expanded JAQL and chart recommendations returned from the chat response.
 *
 * @internal
 */
export interface ExpandedQueryModel {
  /** @internal */
  chartRecommendations: ChartRecommendations | {};
  /** @internal */
  jaql: {
    datasource: {
      id?: string;
      type?: 'elasticube' | 'live';
      title: string;
    };
    metadata: MetadataItem[];
  };
  /** @internal */
  queryTitle: string;
}

/**
 * Empty Expanded Query Model
 *
 * Used as a default value for ExpandedQueryModel
 *
 * @internal
 */
export const EMPTY_EXPANDED_QUERY_MODEL: ExpandedQueryModel = {
  jaql: { datasource: { title: '', type: 'elasticube' }, metadata: [] },
  queryTitle: '',
  chartRecommendations: {},
};

/**
 * Simple Chart Recommendations
 *
 * @internal
 */
export interface SimpleChartRecommendations {
  chartType: string;
  dataOptions: AxesMapping;
  styleOptions: ChartStyleOptions;
}

/**
 * Simple Query Model that is a simplified version of ExpandedQueryModel.
 *
 * It removes any JAQL syntax or information that already exists in the schema.
 * Also, any parameters that have a default do not need to be specified.
 *
 * Users manipulates this model via the query in YAML format.
 *
 * QueryTranslator is responsible for translating this query model to ExpandedWidgetModel.
 *
 * @internal
 */
export interface SimpleQueryModel {
  /**
   * The data model or perspective title.
   */
  model: string;
  metadata: MetadataItem[];
  chart: SimpleChartRecommendations | {};
  /**
   * Title of the query
   */
  queryTitle: string;
}

/**
 * Empty Simple Query Model
 *
 * Used as a default value for SimpleQueryModel
 *
 * @internal
 */
export const EMPTY_SIMPLE_QUERY_MODEL: SimpleQueryModel = {
  model: '',
  metadata: [],
  chart: {},
  queryTitle: '',
};

/**
 * UI Framework
 *
 * @internal
 */
export type UiFramework = 'react' | 'vue' | 'angular';

/**
 * Code Template Key
 *
 * @internal
 */
export type CodeTemplateKey =
  | 'baseChartTmpl'
  | 'chartTmpl'
  | 'chartWidgetTmpl'
  | 'chartWidgetPropsTmpl'
  | 'widgetByIdTmpl'
  | 'executeQueryByWidgetIdTmpl'
  | 'executeQueryTmpl'
  | 'executePivotQueryTmpl'
  | 'pivotTableWidgetTmpl'
  | 'pivotTableWidgetPropsTmpl'
  | 'dashboardByIdTmpl'
  | 'dashboardTmpl';

/**
 * Code Templates
 *
 * @internal
 */
export type CodeTemplates = {
  [key in UiFramework]: { [key in CodeTemplateKey]: string };
};

/**
 * Code Placeholder Map
 *
 * @internal
 */
export type CodePlaceholderMap = Record<string, string>;

/**
 * Base Code Config
 *
 * @internal
 */
export type BaseCodeConfig = { uiFramework?: UiFramework };

/** TYPES FOR WIDGET CODE */

/**
 * Widget Code Config
 *
 * @internal
 */
export type WidgetCodeConfig = BaseCodeConfig & { includeChart?: boolean };

/**
 * Client-side Widget Code Params
 *
 * @internal
 */
export type ClientSideWidgetCodeParams = BaseCodeConfig & {
  widgetProps: WidgetProps;
  removeDefaultProps?: boolean;
};

/**
 * By ID Widget Code Params
 *
 * @internal
 */
export type ByIdWidgetCodeParams = WidgetCodeConfig & {
  dashboardOid: string;
  widgetOid: string;
  chartType: DynamicChartType;
};

/**
 * Widget Code Params
 *
 * @internal
 */
export type WidgetCodeParams = ClientSideWidgetCodeParams | ByIdWidgetCodeParams;

/**
 * Check if widget code params is for client-side code
 *
 * @internal
 */
export const isClientSideWidgetCodeParams = (
  params: WidgetCodeParams,
): params is ClientSideWidgetCodeParams => {
  return 'widgetProps' in params;
};

/**
 * Check if widget code params is for by ID code
 *
 * @internal
 */
export const isByIdWidgetCodeParams = (
  params: WidgetCodeParams,
): params is ByIdWidgetCodeParams => {
  return 'dashboardOid' in params && 'widgetOid' in params;
};

/**
 * Widget Props Config
 *
 * @internal
 */
export type WidgetPropsConfig = { useCustomizedStyleOptions?: boolean };

/**
 * Execute Query Code Params
 *
 * @internal
 */
export type ExecuteQueryCodeParams = BaseCodeConfig & {
  queryParams: ExecuteQueryParams;
};

/**
 * Execute Pivot Query Code Params
 *
 * @internal
 */
export type ExecutePivotQueryCodeParams = BaseCodeConfig & {
  pivotQueryParams: ExecutePivotQueryParams;
};

/**
 * Execute Query Code Props
 *
 * @internal
 */
export type ExecuteQueryCodeProps = Stringify<ExecuteQueryParams> & { extraImportsString: string };

/**
 * @internal
 */
export type ExecutePivotQueryCodeProps = Stringify<ExecutePivotQueryParams> & {
  extraImportsString: string;
};

/**
 * Chart Widget Code Props
 *
 * @internal
 */
export type ChartWidgetCodeProps = Stringify<ChartWidgetProps> & ExtraWidgetCodeProps;

/**
 * Pivot Table Widget Code Props
 *
 * @internal
 */
export type PivotTableWidgetCodeProps = Stringify<PivotTableWidgetProps> & ExtraWidgetCodeProps;

/**
 * Template Key by Widget Type
 *
 * @internal
 */
export type TemplateKeyMapByWidgetType = {
  chart: CodeTemplateKey;
  pivot: CodeTemplateKey;
  text?: CodeTemplateKey;
  custom?: CodeTemplateKey;
};

/** TYPE FOR DASHBOARD CODE */

/**
 * Dashboard Code Config
 *
 * @internal
 */
export type DashboardCodeConfig = BaseCodeConfig;

/**
 * Client-side Dashboard Code Params
 *
 * @internal
 */
export type ClientSideDashboardCodeParams = BaseCodeConfig & {
  dashboardProps: DashboardProps;
};

/**
 * By ID Dashboard Code Params
 *
 * @internal
 */
export type ByIdDashboardCodeParams = DashboardCodeConfig & {
  dashboardOid: string;
};

/**
 * Dashboard Code Params
 *
 * @internal
 */
export type DashboardCodeParams = ClientSideDashboardCodeParams | ByIdDashboardCodeParams;

/**
 * Check if dashboard code params is for client-side code
 *
 * @internal
 */
export const isClientSideDashboardCodeParams = (
  params: DashboardCodeParams,
): params is ClientSideDashboardCodeParams => {
  return 'dashboardProps' in params;
};

/**
 * Check if dashboard code params is for by ID code
 *
 * @internal
 */
export const isByIdDashboardCodeParams = (
  params: DashboardCodeParams,
): params is ByIdDashboardCodeParams => {
  return 'dashboardOid' in params;
};

/**
 * Dashboard Code Props
 *
 * @internal
 */
export type DashboardCodeProps = Stringify<DashboardProps> & ExtraWidgetCodeProps;

/**
 * @internal
 */
export type NormalizedField = {
  name: string;
  dataType: string;
  expression: string;
  description?: string;
};

/**
 * @internal
 */
export type NormalizedTable = {
  name: string;
  columns: NormalizedField[];
};

/**
 * NLQ generic input with context for translation
 *
 * @internal
 */
export type NlqTranslationInput<DataType, ContextType> = {
  data: DataType;
  context: ContextType;
};

/**
 * Response JSON generated by the new NLQ API.
 *
 * @internal
 */
export type NlqResponseJSON = {
  dimensions: JSONArray;
  measures: JSONArray;
  filters: JSONArray;
  highlights?: JSONArray;
};

/**
 * Data schema context for translation
 *
 * @internal
 */
export interface DataSchemaContext {
  /** The data source being used for the query */
  dataSource: JaqlDataSourceForDto;
  /** Available tables for attribute resolution */
  tables: NormalizedTable[];
}

/**
 * Query input for translation
 *
 * @internal
 */
export type QueryInput = NlqTranslationInput<NlqResponseJSON, DataSchemaContext>;

/**
 * Context information for error generation
 *
 * @internal
 */
export type NlqTranslationErrorContext = {
  /** The category of the query element that failed (dimensions, measures, filters, highlights) */
  category: 'dimensions' | 'measures' | 'filters' | 'highlights';
  /** The index of the element within its category array */
  index: number;
  /** The complete original input that caused the error */
  input: JSONValue;
};

/**
 * Structured error information for translation failures.
 *
 * @internal
 */
export type NlqTranslationError = NlqTranslationErrorContext & {
  /** Human-readable error message describing what went wrong */
  message: string;
};

/**
 * Structured error response format for translation failures.
 *
 * @internal
 */
export type NlqTranslationErrorResult = {
  success: false;
  errors: NlqTranslationError[];
};

/**
 * Success response format for translation operations.
 *
 * @internal
 */
export type NlqTranslationSuccessResult<T> = {
  success: true;
  data: T;
};
/**
 * Represents the result of a translation operation that can either succeed with data or fail with errors.
 *
 * @internal
 */
export type NlqTranslationResult<T> = NlqTranslationSuccessResult<T> | NlqTranslationErrorResult;
