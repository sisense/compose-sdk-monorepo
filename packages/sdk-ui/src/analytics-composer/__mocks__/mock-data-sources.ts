import { JaqlDataSourceForDto } from '@sisense/sdk-data';

import { NormalizedTable } from '../types';

export const MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE: JaqlDataSourceForDto = {
  title: 'Sample ECommerce',
  address: 'LocalHost',
  id: 'localhost_aSampleIAAaECommerce',
  live: false,
};
export const MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE: NormalizedTable[] = [
  {
    name: 'Country',
    columns: [
      {
        name: 'Country',
        dataType: 'string',
        expression: '[Country.Country]',
        description: 'Country',
      },
    ],
  },
  {
    name: 'Brand',
    columns: [
      {
        name: 'Brand',
        dataType: 'string',
        expression: '[Brand.Brand]',
        description: 'Brand',
      },
    ],
  },
  {
    name: 'Category',
    columns: [
      {
        name: 'Category',
        dataType: 'string',
        expression: '[Category.Category]',
        description: 'Category',
      },
    ],
  },
  {
    name: 'Commerce',
    columns: [
      {
        name: 'Date',
        dataType: 'datetime',
        expression: '[Commerce.Date]',
        description: 'Date',
      },
      {
        name: 'Revenue',
        dataType: 'number',
        expression: '[Commerce.Revenue]',
        description: 'Revenue',
      },
      {
        name: 'Cost',
        dataType: 'number',
        expression: '[Commerce.Cost]',
        description: 'Cost',
      },
    ],
  },
];
