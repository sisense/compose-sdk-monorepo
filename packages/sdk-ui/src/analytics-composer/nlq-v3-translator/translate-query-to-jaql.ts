import { type DataSource, getFilterListAndRelationsJaql } from '@sisense/sdk-data';
import { getJaqlQueryPayload } from '@sisense/sdk-query-client';
import type { JaqlQueryPayload } from '@sisense/sdk-query-client';

import { BaseQueryParams } from '@/query-execution/types.js';
import { prepareQueryParams, type QueryDescription } from '@/query/execute-query.js';

/**
 * Converts BaseQueryParams to QueryDescription format
 *
 * @param queryParams - The query parameters to convert
 * @returns QueryDescription
 */
function baseQueryParamsToQueryDescription(queryParams: BaseQueryParams): QueryDescription {
  // Extract filters and filter relations from BaseQueryParams
  // BaseQueryParams.filters can be Filter[] | FilterRelations, but QueryDescription.filters is Filter[]
  const { filters: pureFilters, relations: filterRelations } = getFilterListAndRelationsJaql(
    queryParams.filters || [],
  );

  return {
    dataSource: queryParams.dataSource,
    dimensions: queryParams.dimensions,
    measures: queryParams.measures,
    filters: pureFilters,
    filterRelations,
    highlights: queryParams.highlights,
    // No count/offset/ungroup - these are execution concerns, not translation concerns
  };
}

/**
 * Converts BaseQueryParams to JAQL payload
 *
 * This is a pure TypeScript function that works in Node.js environments
 * and converts query parameters to JAQL without requiring React hooks or browser APIs.
 *
 * Reuses the existing query preparation logic from executeQuery to avoid duplication.
 *
 * @param queryParams - The base query parameters from NLQ translation
 * @param defaultDataSource - Default data source if not specified in queryParams
 * @param shouldSkipHighlightsWithoutAttributes - Whether to skip highlights without corresponding attributes
 * @returns The JAQL query payload
 * @internal
 */
export function translateQueryToJaql(
  queryParams: BaseQueryParams,
  defaultDataSource?: DataSource,
  shouldSkipHighlightsWithoutAttributes = false,
): JaqlQueryPayload {
  // Convert BaseQueryParams to QueryDescription format
  const queryDescription = baseQueryParamsToQueryDescription(queryParams);

  // Reuse existing logic from executeQuery to prepare internal query description
  const internalQueryDescription = prepareQueryParams(queryDescription, defaultDataSource);

  // Generate JAQL payload
  return getJaqlQueryPayload(internalQueryDescription, shouldSkipHighlightsWithoutAttributes);
}
