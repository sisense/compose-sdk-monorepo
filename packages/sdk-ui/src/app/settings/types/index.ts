export type { GlobalsObject } from './globals.js';
