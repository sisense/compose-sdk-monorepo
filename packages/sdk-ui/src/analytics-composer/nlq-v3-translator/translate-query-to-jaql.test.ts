import { createAttribute, createMeasure, filterFactory } from '@sisense/sdk-data';
import { describe, expect, it } from 'vitest';

import { BaseQueryParams } from '@/query-execution/types.js';

import { MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE } from '../__mocks__/mock-data-sources.js';
import { translateQueryToJaql } from './translate-query-to-jaql.js';

describe('translateQueryParamsToJaql', () => {
  it('should convert BaseQueryParams to JAQL payload', () => {
    const categoryAttribute = createAttribute({
      name: 'Category',
      type: 'text-attribute',
      expression: '[Category.Category]',
    });
    const categoryCountMeasure = createMeasure({
      name: 'Category Count',
      agg: 'count',
      attribute: categoryAttribute,
    });
    const queryParams: BaseQueryParams = {
      dataSource: {
        ...MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        type: 'elasticube',
      },
      dimensions: [categoryAttribute],
      measures: [categoryCountMeasure],
      filters: [filterFactory.members(categoryAttribute, ['Category A', 'Category B'])],
      highlights: [],
    };

    const result = translateQueryToJaql(queryParams);

    expect(result).toBeDefined();
    expect(result.metadata).toMatchSnapshot();
  });
});
