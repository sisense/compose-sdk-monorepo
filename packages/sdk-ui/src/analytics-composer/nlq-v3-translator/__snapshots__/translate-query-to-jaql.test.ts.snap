// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`translateQueryParamsToJaql > should convert BaseQueryParams to JAQL payload 1`] = `
[
  {
    "jaql": {
      "datatype": "text",
      "dim": "[Category.Category]",
      "title": "Category",
    },
  },
  {
    "jaql": {
      "agg": "countduplicates",
      "datatype": "text",
      "dim": "[Category.Category]",
      "sort": "desc",
      "title": "Category Count",
    },
  },
  {
    "jaql": {
      "datatype": "text",
      "dim": "[Category.Category]",
      "filter": {
        "members": [
          "Category A",
          "Category B",
        ],
      },
      "title": "Category",
    },
    "panel": "scope",
  },
]
`;
